### ISSUE TITLE
Documentation is JSDoc format

### ISSUE THREAD

**Issue by j-martyn** [tshemsedinov/impress#296]
Convert present documentation to JSDoc, and cover functions without documentation.
Features JSDoc gives:
 * [Validating JavaScript code with JSDoc types annotations](http://blog.jetbrains.com/webide/2012/10/validating-javascript-code-with-jsdoc-types-annotations/) (and also [smart type autocomplete](https://www.jetbrains.com/webstorm/help/smart-type-code-completion-completing-code-based-on-type-information.html) etc.). Useful for those who writes code not in FAR :)
 * [generating documentation](http://usejsdoc.org/about-getting-started.html#adding-documentation-comments-to-your-code).

More info:
 *  http://usejsdoc.org/
 * https://github.com/jsdoc3/jsdoc
 * [Type expressions](https://developers.google.com/closure/compiler/docs/js-for-compiler#types)

Reference with https://github.com/tshemsedinov/impress/issues/286.

**Comment by tshemsedinov** [comment]
А ты попробуй, там много примесей и непрямого наследования, IDE все равно не будут понимать что откуда берется. Нужно экспериментировать, возможно, это как-то все можно отконфигурировать или описывать не в коде, не возле функций, а в каких-то специальных файлах заголовком, я не знаю... ну чтобы IDE их парсать умели.

**Comment by tshemsedinov** [comment]
В исходники Impress JSDoc комменты добавлять точно не будем. Ести ли возможность делать аналог файлов заголовков  в формате JSDoc (вот как в C++). И посмотри еще как документирован сам Node.js (например https://github.com/joyent/node/blob/master/lib/net.js и другие основные модули) и как у них генерируются доки. Как комментарии сделаны в хороших модулях типа https://github.com/caolan/async и https://github.com/mongodb/node-mongodb-native

**Comment by j-martyn** [comment]
Из указаных примеров, JSDoc используется только в node-mongodb-native, но у них прямо в коде используется. Но как у кого сделано я изучу.
Сейчас есть такая штука [DefinitelyTyped](http://definitelytyped.org/), там создают что-то вроде header-файлов, в которых описываются заголовки методов с указанием типов и документацией, который потом подключаются комментарием вроде такого 
`/// <reference path="`[`jquery.d.ts`](https://github.com/borisyankov/DefinitelyTyped/blob/master/jquery/jquery.d.ts)`" />`
и IDE их отлично кушает.
Возможно получится что-то такое сделать.

**Comment by tshemsedinov** [comment]
Да, это будет гораздо лучше, кстати, есть вот такая штука http://inch-ci.org/ так она парсит и понимает мои комменты: http://inch-ci.org/github/tshemsedinov/impress

**Comment by j-martyn** [comment]
Твои комменты - это тоже документация, которая очень много где используется, в тех же нодовских исходниках, например. И IDE ее видят, но максимум что они с ней делают - это соотносят с функцией, и отображают документацию как текст. 
В JSDoc же предусмотрен специальный синтаксис, из которого IDE может почерпнуть много полезной инфы, из которых самая полезная - это типизация, как по мне. Их inch-ci тоже будет понимать.

**Comment by tshemsedinov** [comment]
Для документирования я сделал ветку ```doc``` https://github.com/tshemsedinov/impress/tree/doc
В ней есть ```/doc/ru``` и  ```/doc/en``` где я буду описывать свойства и методы объектов.
Кроме того, добавил такой обработчик ```/examples/simple/listSandboxNamespaces.ajsx/``` который возвращает эти же списки свойств и методов объектов из живого приложения.

**Comment by tshemsedinov** [comment]
Кстати, нет необходимости делать, чтобы IDE шарило автодополнение в коде самого Impress, главное, чтобы оно работало в файлах проекта, например, в ```/init/*.js```, в ```/server/*/*.js```, в ```/config/*.js``` и т.д.

**Comment by j-martyn** [comment]
Ну да, по сути надо что бы не внутренности были задокументированы, а сами интерфейсы.
Тут действительно куча всяких подводных камней и непонятностей, пока пробую и экспериментирую, как оно будет лучше работать. А то что в ветке doc, думаю, очень пригодится.

**Comment by tshemsedinov** [comment]
То, что в ветке doc нужно перегнать в какой-то удобоваримый формат, из которого умеет читать IDE или же в JSON, чтобы потом из него автоматически транслировать в формат для IDE и в HTML для просмотра на сайте в качестве доков.
